smolyay/basis.py

```diff
@@ -403,29 +482,393 @@ def _derivative(self, x):
         float
             Value of the derivative of Trigonometric polynomial.
 
+        n : int, optional
+            order of derivative. Default is 1.
+
         Raises
         ------
         ValueError
             If input is outside the domain `[0, 2\pi]`
         """
         x = numpy.asarray(x)
-        return self.frequency * 1j * numpy.exp(x * self.frequency * 1j)
+        return numpy.exp(x * self.frequency * 1j) * (self.frequency * 1j) ** n
+
+
+class BasisFunctionSet(collections.abc.Sequence):
+    """Set of basis functions and sample points."""
+
+    @abc.abstractmethod
+    def __init__(self):
+        pass
+
+    @property
```

Comment:
Add a setter for this property that checks if all the basis functions have the same type and domain, and errors if not.

Reply:
Wouldn't that make the property mutable to the user? This would allow `basis_functions` to be changed arbitrarily even if the class is supposed to be a specific type of basis function.

Reply:
That's true—you will need to do this in the constructor then.

Reply:
I understand why basis functions should be limited to the same domain, but why type? Say a hypothetical wants to use Chebyshev polynomials to model some function but also wants to throw in an extra exponential decay term. As long as that basis function uses the same domain, I see no reason why that should be banned outright. All the terms in `basis_functions` should derive from `BasisFunction`, but that's all that should matter in terms of typing.

Reply:
We don't currently have a use case where we mix basis functions within a 1 dimensional set. So, it would be better to be more restrictive now and relax it later if we want to, after we have thought through the implications of it fully.

Reply:
Since `basis_functions` is without a setter in `BasisFunctionSet`, does this give `CustomBasisFunctionSet` a reason to exist to implement that setter?

Reply:
We don't currently have any use cases for making basis function sets other than what we have, so no, I don't think so. A custom list of basis functions could be assigned through the constructor of `BasisFunctionSet`.

Reply:
okay!